notebooks/lesson_preprocessor.py

```diff
@@ -1,74 +1,91 @@
     logging.basicConfig(level=logging.DEBUG)
 
+IGNORE_UNKNOWN_MACROS = 0
 
-def load_lessons(lesson_dicts):
-    res = [Lesson(d, i) for i, d in enumerate(lesson_dicts)]
-    for i, lesson in enumerate(res):
-        if i > 0:
-            lesson._prev = res[i-1]
-            lesson.first = False
-        else:
-            lesson.first = True
-        if i < len(res)-1:
-            lesson._next = res[i+1]
-            lesson.last = False
-        else:
-            lesson.last = True
-    return res
-
-class Lesson:
-    def __init__(self, d, idx):
-        self.num = idx+1
-        for k, v in d.items():
-            if isinstance(v, dict):
-                v = Box(v)
-            setattr(self, k, v)
-
-    @property
-    def exercise_forking_url(self):
-        return 'https://www.kaggle.com/kernels/fork/{}'.format(self.exercise.scriptid)
-
-    @property
-    def tutorial_url(self):
-        return 'https://www.kaggle.com/{}'.format(self.tutorial.slug)
-
-    @property
-    def next(self):
-        if self.last:
-            raise ValueError("Lesson number {} is the last. No next lesson.".format(self.num))
-        return self._next
-    @property
-    def prev(self):
-        if self.first:
-            raise ValueError("Lesson number {} is the first. No prev lesson.".format(self.num))
-        return self._prev
+class UnrecognizedMacroException(Exception):
+    pass
 
-class LearnLessonPreprocessor(Preprocessor):
+def get_git_branch():
+    """Return the current git branch as a string"""
+    return subprocess.check_output(["git", "symbolic-ref", "--short", "HEAD"])\
+            .decode('utf8').strip()
 
-    lessons_metadata = traitlets.List().tag(config=True)
+class LearnLessonPreprocessor(Preprocessor):
     
+    # NB: This is the only overridden Preprocessor method. All other methods are custom.
     def preprocess(self, nb, resources):
-        lt_meta = nb['metadata']['learntools_metadata']
-        lesson_ix = lt_meta['lesson_index']
-        lessons = load_lessons(self.lessons_metadata)
-        self.lessons = lessons
-        self.lesson = lessons[lesson_ix]
+        # See render.py for how resources as populated.
+        self.track = resources['track_meta']
+        # May be None for notebooks with type='extra'
+        self.lesson = resources['lesson']
+        # Corresponds to track_config.yaml
+        track_cfg = resources['track_cfg']
+        nb_meta = resources['nb_meta']
+
         for i, cell in enumerate(nb.cells):
             nb.cells[i] = self.process_cell(cell)
+        # NB: There may be some cases where we need to access learntools in a tutorial
+        # or ancillary notebook as well. Could encode this in track_meta, or if we wanted
+        # to be really clever, could look for learntools imports in code cells.
+        if track_cfg.get('development', False) and nb_meta.type == 'exercise':
+            self.pip_install_lt_hack(nb)
         return nb, resources
+    
+    def pip_install_lt_hack(self, nb):
```

Comment:
What are the use cases for these pip install hacks?

I haven't seen how these are used yet, but I'm very concerned about us pip installing things in kernels... That applies to both learntools and everything else.

The behavior of `pip installation` in kernels is likely to change frequently in the near future.  

If we're installing things other than learntools, we should say what we're doing in the markdown.  Once you do that, you might as well right the `pip install` line yourself.

Regarding installing learntools like this, it seems somewhat implicit, and I'm concered that increases the risk we'll accidentally leave it in when we ought to be replacing it.

Reply:
This is only used when cfg.development is True (so it's not intended for published, user-facing  notebooks). 

It's basically an alternative to zipping up your learntools tree and adding it as a data source. I've been using it for testing the embeddings course.

I'm not worried about the changes to how pip install works on Kernels breaking this, because kernels using this trick are ephemeral by nature. I definitely look forward to simplifying this when b/111888649 is fixed.

Reply:
Do you suggest turning this off during a challenge, going back to datasets + path munging?  Or keeping it on as a replacement for our previous dataset approach?

Reply:
For challenges, I've made changes during the day after I released a lesson. I don't think the main docker image is going to be fast enough to allow that.

That makes me think we have a few options
1. Use this for the challenge (I agree with you that it's uglier than the dataset hack)
2. Do the dataset hack manually (if we do that, I think we'll want to use the dataset hack rather than this pip_hack for testing on Kaggle as well.)
3. Do the dataset hack programmatically... as in create something akin to the programmatic pip hack, but have it push the dataset and munge the path. I suspect this is more complex than 2, since it involves both pushing the dataset and munging the notebook contents.

Seems fine to me to leave this like it is, and verify this challenge thing is a real use case by going through it a couple more times (using option 1 or 2 above, which don't require more code yet).

Reply:
+1 to all that

> Do the dataset hack manually (if we do that, I think we'll want to use the dataset hack rather than this pip_hack for testing on Kaggle as well.)

Though I'd still prefer to use the pip trick for testing if I'm working on something like the embeddings course that's not intended to be launched as a challenge.

Also, not a bridge we need to cross yet, but I do think that once we find ourselves working on another challenge, 3 will pay for itself pretty quickly (in terms of net time compared to 2). I actually started writing a bit of automation around the dataset trick a month or two ago, but quickly dropped it because I was hitting too many datasets API bugs. The tickets I filed have since been fixed (and the API in general feels a lot more stable/usable), so I don't think it would be a tall task.